### PR TITLE
Ssh poll

### PR DESCRIPTION
**Pull request by mfriedl**

### CODE REVIEW

**Comment by mfriedl** [comment]
@djmdjm i've addressed your comments

File: usr.bin/ssh/sshd.c [hunk 1]
```diff
@@ -939,15 +935,12 @@ server_accept_loop(int *sock_in, int *sock_out, int *newsock, int *config_s,
 			startup_pollfd[i] = -1;
 			if (children[i].pipefd != -1) {
 				pfd[npfd].fd = children[i].pipefd;
-				pfd[npfd].events = POLLIN;
+				if (children[i].config != NULL)
```
Comment (djmdjm) [inline usr.bin/ssh/sshd.c]: Maybe set POLLIN unconditionally? This could give better diagnostics if a sshd-session exited before it finishes reading the config.

If we do this then I think there needs to be some tests that `child->config == NULL` after the read() below - specifically the child should never write anything to the configfd until the config is completely sent.

File: usr.bin/ssh/sshd.c [hunk 2]
```diff
@@ -293,9 +289,9 @@ child_close(struct early_child *child, int force_final, int quiet)
 		close(child->pipefd);
 		child->pipefd = -1;
 	}
-	if (child->configfd != -1) {
-		close(child->configfd);
-		child->configfd = -1;
+	if (child->config != NULL) {
```
Comment (djmdjm) [inline usr.bin/ssh/sshd.c]: It shouldn't matter but move this to `child_finish()` because that does the other free-like stuff

It can also be unconditional - `sshbuf_free(NULL)` is fine

File: usr.bin/ssh/sshd.c [hunk 3]
```diff
@@ -725,14 +720,16 @@ pack_config(struct sshbuf *conf)
...
+	struct sshbuf *m;
 	u_int mlen;
 	int sz;
 
-	debug3_f("entering fd = %d rexec_state len %zu", fd,
-	    sshbuf_len(m));
+	debug3_f("entering fd = %d config len %zu", fd,
+	    sshbuf_len(config));
 
+	m = pack_config(config);
```
Comment (djmdjm) [inline usr.bin/ssh/sshd.c]: We could pack the config once and only serialise/send the keys after the config blob has been sent. E.g. 

1. Make a global `sshbuf *config_blob`
2. Fill `child->config` using `sshbuf_fromb(config_blob)`, so it's zero-copy
3. Add a new `child->keys` that only gets filled in and sent once `child->config` has finished sending

The benefitd to this are 1) less work for the listener it doesn't have to serialise the config each time, 2) less memory as it doesn't have to hold on to `max_startup * sshbuf_len(child->config)` in the worst case, and 3) the keys are exposed in memory for much less time (I expect that in the usual case they would instantly fit in the socket buffer).

OTOH it's a fair bit of work

File: usr.bin/ssh/sshd.c [hunk 4]
```diff
@@ -962,33 +955,28 @@ server_accept_loop(int *sock_in, int *sock_out, int *newsock, int *config_s,
...
-				goto config_send_failed;
 			if (ret == len) {
 				/* finished sending config */
-				close(children[i].configfd);
-				children[i].configfd = -1;
+				sshbuf_free(children[i].config);
+				children[i].config = NULL;
 			} else {
-				children[i].sent += ret;
+				sshbuf_consume(children[i].config, ret);
```
Comment (djmdjm) [inline usr.bin/ssh/sshd.c]: should check the return status from sshbuf_consume() here :)

**Review by djmdjm — COMMENTED** [review]
Looks good! Some minor feedback enclosed.

Github won't let me comment on unchanged lines, but I think the logging in `child_reap()` needs to change too because we're adding another state to the child's lifecycle. ATM it logs stuff like this:

```C
		error("session process %ld for %s terminated abnormally, "
		    "status=0x%x%s", (long)child->pid, child->id, child->status,
		    child->early ? " (early)" : "");
```

I think we should replace the `child->early ? " (early)" : ""` with a status string that includes the "still sending config" state too.

File: usr.bin/ssh/sshd.c (hunk 1 above)
Comment (mfriedl) [reply usr.bin/ssh/sshd.c]: done

File: usr.bin/ssh/sshd.c (hunk 2 above)
Comment (mfriedl) [reply usr.bin/ssh/sshd.c]: done

File: usr.bin/ssh/sshd.c (hunk 4 above)
Comment (mfriedl) [reply usr.bin/ssh/sshd.c]: done

File: usr.bin/ssh/sshd.c (hunk 3 above)
Comment (mfriedl) [reply usr.bin/ssh/sshd.c]: i'll have a look

Comment (mfriedl) [reply usr.bin/ssh/sshd.c]: done

File: usr.bin/ssh/sshd.c [hunk 5]
```diff
@@ -323,6 +320,14 @@ child_reap(struct early_child *child)
 {
 	LogLevel level = SYSLOG_LEVEL_DEBUG1;
 	int was_crash, penalty_type = SRCLIMIT_PENALTY_NONE;
+	const char *child_status;
```
Comment (djmdjm) [inline usr.bin/ssh/sshd.c:326]: set `child_status = ""` here and save two lines below :)

File: usr.bin/ssh/sshd.c [hunk 6]
```diff
@@ -1000,6 +1008,15 @@ server_accept_loop(int *sock_in, int *sock_out, int *newsock, int *config_s,
 				child_close(&(children[i]), 0, 0);
 				break;
 			case 1:
+				if (children[i].config) {
+					error_f("startup pipe %d (fd=%d)"
+					    "early read", i, children[i].pipefd);
+					if (children[i].early)
+						listening--;
+					srclimit_done(children[i].pipefd);
+					child_close(&(children[i]), 0, 0);
```
Comment (djmdjm) [inline usr.bin/ssh/sshd.c:1064]: maybe even kill the child here? I don't know...

File: usr.bin/ssh/sshd-session.c [hunk 7]
```diff
@@ -648,7 +650,6 @@ recv_rexec_state(int fd, struct sshbuf *conf, uint64_t *timing_secretp)
 		fatal_f("rexec version mismatch");
 	if ((r = sshbuf_get_string(m, &cp, &len)) != 0 || /* XXX _direct */
 	    (r = sshbuf_get_u64(m, timing_secretp)) != 0 ||
-	    (r = sshbuf_froms(m, &hostkeys)) != 0 ||
```
Comment (djmdjm) [inline usr.bin/ssh/sshd-session.c:652]: I didn't consider this before when I suggested it, but this makes this function non-symmetric with the sshd-session->sshd-auth config passing...

File: usr.bin/ssh/sshd.c [hunk 8]
```diff
@@ -718,34 +716,57 @@ pack_config(struct sshbuf *conf)
...
+ *	uint8   type (ignored)
+ *	string	host_keys[] {
+ *		string private_key
+ *		string public_key
+ *		string certificate
+ *	}
+ */
 static void
-send_rexec_state(int fd, struct sshbuf *config)
+send_rexec_state(int fd)
```
Comment (djmdjm) [inline usr.bin/ssh/sshd.c:751]: Maybe add a comment here that this function is only called for `debug_mode || inetd_mode` and that the usual case is handled asynchronously by `server_accept_loop()`

File: usr.bin/ssh/sshd.c (hunk 5 above)
Comment (mfriedl) [reply usr.bin/ssh/sshd.c:326]: i wanted to have the setting aligned, especiially since there is a 2nd case in a for-loop where this would be wrong

File: usr.bin/ssh/sshd-session.c (hunk 7 above)
Comment (mfriedl) [reply usr.bin/ssh/sshd-session.c:652]: i can skip this change if you think it's not improving things

File: usr.bin/ssh/sshd.c (hunk 6 above)
Comment (mfriedl) [reply usr.bin/ssh/sshd.c:1064]: i've added a kill SIGTERM

File: usr.bin/ssh/sshd.c (hunk 8 above)
Comment (mfriedl) [reply usr.bin/ssh/sshd.c:751]: I've added a comment, but also changed to function to use fork() so the config is sent from the  child proc to the exec-ed parent

this avoids the problems w/ the socketpair buffer size

we've been talking about this before, but I think we could add the fork() here (it's only debug/inetd mode)

File: usr.bin/ssh/sshd-session.c (hunk 7 above)
Comment (djmdjm) [reply usr.bin/ssh/sshd-session.c:652]: I think it's fine, just noting it in case we want to add a comment elsewhere noting the divergence

File: usr.bin/ssh/sshd.c (hunk 8 above)
Comment (djmdjm) [reply usr.bin/ssh/sshd.c:751]: IMO that's fine

**Comment by djmdjm** [comment]
Looks good to me!
